sample_project/Source/sample_project/Measure/Measure.cpp

```diff
@@ -0,0 +1,261 @@
+	if (WidgetAsset.Succeeded()) {
+		UIWidgetClass = WidgetAsset.Object;
+	}
+
+	static ConstructorHelpers::FObjectFinder<UStaticMesh> MeshAsset(TEXT("/Game/SampleViewer/Samples/Routing/Geometries/Cube.Cube"));
+	if (MeshAsset.Succeeded()) {
+		RouteMesh = MeshAsset.Object;
+	}
+
+}
+
+// Called when the game starts or when spawned
+void AMeasure::BeginPlay()
+{
+	Super::BeginPlay();
+
+	SetupInput();
+
+	// Make sure mouse cursor remains visible
+	APlayerController* PC = UGameplayStatics::GetPlayerController(GetWorld(), 0);
+	if (PC)
+	{
+		PC->bShowMouseCursor = true;
+		PC->bEnableClickEvents = true;
+	}
+
+	// Create the UI and add it to the viewport
+	if (UIWidgetClass)
+	{
+		UIWidget = CreateWidget<UUserWidget>(GetWorld(), UIWidgetClass);
+		if (UIWidget)
+		{
+			UIWidget->AddToViewport();
+			WidgetFunction = UIWidget->FindFunction(FName("SetDistance"));
+			UnitDropdown = (UComboBoxString*)UIWidget->GetWidgetFromName(TEXT("UnitDropDown"));
+		}
+	}
+	unitTxt = " m";
+	unit = UArcGISLinearUnit::CreateArcGISLinearUnit(EArcGISLinearUnitId::Meters);
+	currentUnit = m;
+}
+
+// Called every frame
+void AMeasure::Tick(float DeltaTime)
+{
+	Super::Tick(DeltaTime);
+
+}
+
+void AMeasure::SetupInput()
+{
+	InputComponent = NewObject<UInputComponent>(this);
+	InputComponent->RegisterComponent();
+
+	if (InputComponent)
+	{
+		InputComponent->BindAction("PlaceRoutePoint", IE_Pressed, this, &AMeasure::AddStop);
+		EnableInput(GetWorld()->GetFirstPlayerController());
+	}
+}
+void AMeasure::AddStop()
+{
+	float traceLength = 10000000.f;
+	FVector position;
+	FVector direction;
+	FHitResult hit;
+	FVector tangent;
+
+	APlayerController* PlayerController = UGameplayStatics::GetPlayerController(GetWorld(), 0);
+	PlayerController->DeprojectMousePositionToWorld(position, direction);
+
+	bool bTraceSuccess = GetWorld()->LineTraceSingleByChannel(hit, position,
+		position + traceLength * direction, ECC_Visibility, FCollisionQueryParams());
+
+	if (bTraceSuccess && hit.GetActor()->GetClass() == AArcGISMapActor::StaticClass() && hit.bBlockingHit)
+	{
+		FActorSpawnParameters SpawnParam = FActorSpawnParameters();
+		SpawnParam.Owner = this;
+
+		auto lineMarker = GetWorld()->SpawnActor<ARouteMarker>(ARouteMarker::StaticClass(), hit.ImpactPoint, FRotator(0.f), SpawnParam);
+
+		auto thisPoint = lineMarker->ArcGISLocation->GetPosition();
+
+		if (!stops.IsEmpty())
+		{
+			auto lastStop = stops.Last();
+			auto lastPoint = lastStop->ArcGISLocation->GetPosition();
+
+			//calculate distance from last point to this point
+			double d = UArcGISGeometryEngine::DistanceGeodetic(lastPoint, thisPoint, unit, UArcGISAngularUnit::CreateArcGISAngularUnit(EArcGISAngularUnitId::Degrees), EArcGISGeodeticCurveType::Geodesic)->GetDistance();
+			geodeticDistance += d;
+			geodeticDistanceText = FString::Printf(TEXT("Distance: %f %s"), round(geodeticDistance * 1000.0) / 1000.0, *unitTxt);
```

Comment:
It might make more sense to make this local a local var with `auto`

```suggestion
			auto geodeticDistanceText = FString::Printf(TEXT("Distance: %f %s"), round(geodeticDistance * 1000.0f) / 1000.0f, *unitTxt);
```

Reply:
this one is used by multiple functions, for example in `unitChanged`, whenever the unit selector detects a unit change geodeticDistanceText gets updated (1 miles -> 1.6 kilometers)